cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,308 @@
+ * the License.
+ */
+
+package co.cask.cdap.metadata;
+
+import co.cask.cdap.api.lineage.field.EndPoint;
+import co.cask.cdap.api.lineage.field.InputField;
+import co.cask.cdap.api.lineage.field.Operation;
+import co.cask.cdap.api.lineage.field.OperationType;
+import co.cask.cdap.api.lineage.field.ReadOperation;
+import co.cask.cdap.api.lineage.field.TransformOperation;
+import co.cask.cdap.api.lineage.field.WriteOperation;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.metadata.lineage.field.DefaultFieldLineageReader;
+import co.cask.cdap.data2.metadata.lineage.field.EndPointField;
+import co.cask.cdap.data2.metadata.lineage.field.FieldLineageReader;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ProgramRunId;
+
+import co.cask.cdap.proto.metadata.lineage.DatasetField;
+import co.cask.cdap.proto.metadata.lineage.FieldLineageDetails;
+import co.cask.cdap.proto.metadata.lineage.FieldLineageSummary;
+import co.cask.cdap.proto.metadata.lineage.FieldOperationInfo;
+import co.cask.cdap.proto.metadata.lineage.FieldOperationInput;
+import co.cask.cdap.proto.metadata.lineage.FieldOperationOutput;
+import co.cask.cdap.proto.metadata.lineage.ProgramFieldOperationInfo;
+import co.cask.cdap.proto.metadata.lineage.ProgramInfo;
+import co.cask.cdap.proto.metadata.lineage.ProgramRunOperations;
+import com.google.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/**
+ * Service to compute field lineage based on operations stored in {@link DefaultFieldLineageReader}.
+ */
+public class FieldLineageAdmin {
+
+  private final FieldLineageReader fieldLineageReader;
+
+  @Inject
+  FieldLineageAdmin(FieldLineageReader fieldLineageReader) {
+    this.fieldLineageReader = fieldLineageReader;
+  }
+
+  /**
+   * Get the set of fields written to the EndPoint by field lineage {@link WriteOperation},
+   * over the given time range, optionally prefixed with the given prefix.
+   *
+   * @param endPoint the EndPoint for which the fields need to be returned
+   * @param start start time (inclusive) in milliseconds
+   * @param end end time (exclusive) in milliseconds
+   * @param prefix prefix for the field name, if {@code null} then all fields are returned
+   * @return set of fields written to a given EndPoint
+   */
+  public Set<String> getFields(EndPoint endPoint, long start, long end, @Nullable String prefix) {
+    Set<String> fields = fieldLineageReader.getFields(endPoint, start, end);
+    if (prefix == null) {
+      return fields;
+    }
+
+    Set<String> prefixFields = new HashSet<>();
```

Comment:
it would be more efficient to push down the filter to the table scan. Can you add a JIRA and TODO?

Reply:
Forgot to post JIRA before - https://issues.cask.co/browse/CDAP-13732